contracts/AelinPool.sol

```diff
@@ -209,49 +212,99 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
+        for (uint256 i; i < nftPurchaseListLength; ++i) {
+            nftPurchaseList = _nftPurchaseList[i];
+            collectionAddress = nftPurchaseList.collectionAddress;
+            tokenIds = nftPurchaseList.tokenIds;
+            tokenIdsLength = tokenIds.length;
+            nftCollectionRules = nftCollectionDetails[collectionAddress];
 
             require(collectionAddress != address(0), "collection should not be null");
-
-            NftCollectionRules memory nftCollectionRules = nftCollectionDetails[collectionAddress];
             require(nftCollectionRules.collectionAddress == collectionAddress, "collection not in the pool");
 
-            if (nftCollectionRules.purchaseAmount > 0) {
-                maxPurchaseTokenAmount += nftCollectionRules.purchaseAmount * tokenIds.length;
-            }
-
-            if (nftCollectionRules.purchaseAmount == 0) {
-                maxPurchaseTokenAmount = _purchaseTokenAmount;
-            }
+            //Iterate over the token ids
+            for (uint256 j; j < tokenIdsLength; ++j) {
+                if (NftCheck.supports721(collectionAddress)) {
+                    require(IERC721(collectionAddress).ownerOf(tokenIds[j]) == msg.sender, "has to be the token owner");
+                    // If there are no ranges then no need to check whether token Id is within them
+                    // Or whether there are any rangeAmounts
+                    if (nftCollectionRules.idRanges.length > 0) {
+                        (bool isTokenIdInRange, uint256 rangeAmountForTokenId) = getRangeData(
+                            tokenIds[j],
+                            nftCollectionRules.idRanges
+                        );
+                        require(isTokenIdInRange, "tokenId not in range");
+
+                        //if there's a range amount for this token id, then set that as the max for its element
+                        if (rangeAmountForTokenId != 0) {
+                            maxPurchaseTokenAmount = incrementMaxPurchaseTokenAmount(
+                                maxPurchaseTokenAmount,
+                                rangeAmountForTokenId
+                            );
+                        } else {
+                            //Otherwise defer to purchaseAmount
+                            if (nftCollectionRules.purchaseAmount == 0) {
+                                maxPurchaseTokenAmount = type(uint256).max;
+                            } else {
+                                maxPurchaseTokenAmount = incrementMaxPurchaseTokenAmount(
+                                    maxPurchaseTokenAmount,
+                                    nftCollectionRules.purchaseAmount
+                                );
+                            }
+                        }
+                    } else {
+                        if (nftCollectionRules.purchaseAmount == 0) {
+                            maxPurchaseTokenAmount = type(uint256).max;
```

Comment:
I think we could really optimise this by breaking the loop if we get one single `type(uint256).max;`. Because why would we continue? No need, we already know the final outcome of this function